### ISSUE TITLE
Unnecessary re-rendering of ComplexHeatmap 

### ISSUE THREAD

**Issue by daymecita** [iSEE/iSEE#580]
Dear iSEE team, 
I am a big fan of iSEE and I use it a lot. That's why I recently found something I think is a bug and compromises the app performance, overall with big datasets.
The ComplexHeatmap is being rendered every time I make a search on the RowDataTable and the heatmap is receiving row selection from the table ("Selection Parameters/Receive row selection from"). This should be OK if the option "Use custom rows" in the Data Parameters box of the heatmap was not selected. If this option is selected, the heatmap shouldn't be re-rendered until I imported the selection ("Edit features names/Import selection" buttons) and applied the changes (Apply button). 
Looking at the code, I see that in the function create_child_propagation_observer, regenerate variable gets TRUE when I do any search on the table, and apparently the children (in this case the ComplexHeatmap) should me modified. 
I'd appreciate you looked into that when you have time. 
Thanks!

**Comment by kevinrue** [comment]
Hi

Thanks for the kind words and the bug report.

I just want to acknowledge that I didn't miss it. I'm in the process of reorganising some of the iSEE code into sub-packages, but once that will be done, I'll look into this issue.

The heatmap has always been a bit of an odd package - especially as we're not using ggplot2 for that panel - so feedback is always useful from the user experience. Thank you!

**Comment by LTLA** [comment]
This is unlikely to be a bug, just a consequence of how the reactivity is architectured for generic panels.

If you set one panel to receive a selection from another panel, it will _always_ trigger a re-rendering of the former when a new selection is made on the latter. This behavior is consistent across all panels, be it a heatmap or scatter plot or table.

The key thing is that this trigger happens before any panel-specific options are applied. In the case of the heatmap, selecting or unselecting the "Use custom rows" option will have no effect on the re-rendering trigger; the re-rendering still occurs. Specifically, selecting the option will wait for a received selection from the upstream panel; ignore that selection; and use the custom row selection instead. Yes, it's a bit of weird that it needs to wait for something that is ignored anyway, but this approach was preferable to building a heatmap-specific message passing system that the rest of the app would have to pay attention to.

So. There are several approaches to avoiding the unnecessary re-rendering:

- On the user side, you can remember to unset the "Receive row selection" when using the "Use custom rows" option. This should break the dependency between panels such that changes to the table do not trigger re-rendering.
- One approach on our side is to alter the "Use custom rows" observer so that it `updateSelectInput`'s the "Receive row selection" input to the default no-selection. This is not 100% well-behaved as different UI elements try not to interfere with each other; in particular, users will have to remember to re-select the upstream panel if they turn off the "Use custom rows".
- I suppose another approach would be to have the "Use custom rows" observer directly modify the reactivity graph, removing or adding the relationship depending on whether it was selected or not. This would require some testing to check for unpleasant side-effects, especially with respect to the dynamic multiple selections. It is also likely that a new generic would need to be introduced to avoid initializing the relationship in `.spawn_multi_selection_graph`.

**Comment by daymecita** [comment]
Hi @kevinrue and @LTLA, 

Many thanks for your replies! 
I understand the transmission between panels works the same for all but maybe the heatmap needs something different. Every character I type on the table Search box triggers a rendering of the heatmap. I am using iSEE with Rsconnect, so it takes longer than when running locally. Sometimes the panel keeps blank for several seconds and no one knows what is happening. 
My other questions are:
1. Why doesn't the heatmap have the loading signal as the other panels have?
2. What happened to the autocomplete functionality when typing gene names that was in previous versions?

**Comment by LTLA** [comment]
> Every character I type on the table Search box triggers a rendering of the heatmap

I would suggest using the first workaround mentioned in my reply above, i.e., set the "Receive row selection from" to "---" in the heatmap's selection parameters.

**Comment by kevinrue** [comment]
Having finally just found some time for testing, I'm actually confused myself: the heatmap is not refreshing at all for me when I:

- receive row selection from Row data table 1
- de-select "use custom rows"

I'm using devel and I've been working on the packages locally, but nothing I've done recently is supposed to impact that part of the code. I'll keep testing ...

**Comment by kevinrue** [comment]
> Every character I type on the table Search box triggers a rendering of the heatmap

One more thought I just had is that we could also alleviate the "every single character" issue by increasing the delay for observing the search box. In other words, if we observe the search box every second or so, users would have more time to type multiple characters before the app would detect the change and trigger a re-rendering.

e.g. <https://shiny.rstudio.com/reference/shiny/1.0.4/debounce.html>

**Comment by kevinrue** [comment]
@daymecita 

I've taken the time to record a [video](https://www.youtube.com/watch?v=3vce4QucXAM&list=PLd1mh2_w6ahZcNyQL-MtACdLUxj2cym5q ) of myself trying to replicate the issue.

Either I cannot reproduce your issue or I misunderstand it.

Can I ask you to watch it and let me know what you think?  Thank you!

**Comment by daymecita** [comment]
Hi Kevin!

I really appreciate the time you took and the video you made very well explained. 
Using your same example I could reproduce the issue by adding a print in the renderPlot observer from the renderOutput function in the panel_ComplexHeatmapPlot.R (see below). With a small dataset like yours no one can notice the re-render, but since I work with more than 30000 cells usually it is more obvious and there are more delays. Please, let me know if I can provide you with more info. 
<img width="479" alt="Screenshot 2022-07-01 194649" src="https://user-images.githubusercontent.com/100194518/176945693-bdbf1eff-61d6-425d-8a69-4e7e4fffc119.png">

**Comment by kevinrue** [comment]
Spinner added in https://github.com/iSEE/iSEE/pull/585
Clearly, I just forgot to add it at the time. It looks like it works just fine for this panel too.
Give it a couple of days to reach the devel branch (iSEE version 2.9.1).

**Comment by daymecita** [comment]
>

**Comment by daymecita** [comment]
Hi Kevin, 

Many thanks for looking into the things I pointed out. The only way I found the heatmap was re-rendering was adding a print statement to the rendering part. I used the latest version and modified only that. Anyways, I will take into account your suggestions for future issues. 
Cheers!

**Comment by kevinrue** [comment]
@daymecita 

I've made some more progress.

Are you able to install the version from the GitHub branch as follows:

```
BiocManager::install("iSEE/iSEE@kra-580-3")
```

I've added some mechanisms that allow the ComplexHeatmapPlot panel to rerender properly in the two scenarios described in my latest message, namely:

- an incoming ROW selection changes and the the checkbox `Use custom rows` is NOT CHECKED.
- an incoming COLUMN selection changes and the checkbox `Restrict to selected columns` is CHECKED.

Once that clears the automated checks and peer-review by fellow iSEE developers, it should finally solve your issue entirely :)

**Comment by kevinrue** [comment]
I've fixed a few glitches that the PR created with respect to existing propagation through chains of panels.

Here's my latest test script:
https://github.com/iSEE/iSEEdebug/blob/main/issues/iSEE-issue-580/script_05.R

And the latest version of the PR can be installed as described above:
```R
BiocManager::install("iSEE/iSEE@kra-580-3")
```

Let me know what you think

**Comment by kevinrue** [comment]
closed by #593 

let me know if iSEE version `2.9.5` re-renders the heatmap in every way you hoped :)

**Comment by daymecita** [comment]
Hi @kevinrue, I just tested the latest iSEE version 2.9.5, and it works exactly how I liked. I am very happy with the results, the heatmap runs smoother now. Super thanks!
